### PR TITLE
Container model for complete data set

### LINKED ISSUE #59: Container model for complete data set
... that is able to generate a structure of single grid models or aggregate different single grid models to a joint data set.

### PR DESCRIPTION
**Pull request by ckittl**
Resolves #59

### CODE REVIEW

**Comment by ckittl** [comment]
!test

**Comment by johanneshiry** [comment]
!test

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+            .collect(Collectors.toSet());
+
+    Set<MeasurementUnitInput> measurements =
+        input.getMeasurementUnits().stream()
+            .filter(measurement -> measurement.getNode().getSubnet() == subnet)
+            .collect(Collectors.toSet());
+
+    return new RawGridElements(nodes, lines, transformer2w, transformer3w, switches, measurements);
+  }
+
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java]:
```suggestion
  private static <T extends SystemParticipantInput> Set<T> filterParticipants(
      Set<T> systemParticipantInputs, int subnet) {
    return systemParticipantInputs.stream()
        .filter(entity -> entity.getNode().getSubnet() == subnet)
        .collect(Collectors.toSet());
  }
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+    return new SystemParticipants(
+        bmPlants,
+        chpPlants,
+        new HashSet<>(),
+        fixedFeedIns,
+        heatpumps,
+        loads,
+        pvs,
+        storages,
+        wecPlants);
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java]:
```suggestion
       Set<BmInput> bmPlants = filterParticipants(input.getBmPlants(), subnet);

    Set<ChpInput> chpPlants = filterParticipants(input.getChpPlants(), subnet);

    /* Electric vehicle charging systems are currently dummy implementations without nodal reverence */
    Set<FixedFeedInInput> fixedFeedIns = filterParticipants(input.getFixedFeedIns(), subnet);

    Set<HpInput> heatpumps = filterParticipants(input.getHeatPumps(), subnet);

    Set<LoadInput> loads = filterParticipants(input.getLoads(), subnet);

    Set<PvInput> pvs = filterParticipants(input.getPvPlants(), subnet);

    Set<StorageInput> storages = filterParticipants(input.getStorages(), subnet);

    Set<WecInput> wecPlants = filterParticipants(input.getWecPlants(), subnet);

    return new SystemParticipants(
        bmPlants,
        chpPlants,
        new HashSet<>(),
        fixedFeedIns,
        heatpumps,
        loads,
        pvs,
        storages,
        wecPlants);
```

File: docs/test/complexTopology.drawio [hunk 3]
```diff
@@ -0,0 +1 @@
...RDcaj5501ou/PvNW1HW44OOw7g05QcsPRaw9n+/LZJ5IEtBpsFnHelzMozrU8llYigdYFbpuKBDsX1NazSJy+vPli64z1ItQWhnzgWq/Har2+QZDsii4a9NRFI71EII5ZxblrYxHYakXAxVtSQRT7ZOqNoBWysNBLQo8W7Oh5+xIR9ulAvnYILRrf/pIU1D64FJJ52i60dFJ0PzC+FoZkNTENa8jhNZDnygfaXs7QBuVrh7brjmkTR2jAUs2bPD8/8DY7wxuSr8H7pq1YvdCGtlsMCRhCzwLlsPegIQHQJCagHYhOFUSsGsQmk2rdQHSqCwLqldhkwqsdiML0T7kSm0wINQMRAN2U2GQSph2ImikRynMY7UFEpq2XEosq3BSI1UQY9UqUp2Xag4iBsNqgXIk3uIaJQXXGol6JTWYs97T4gq4NwxaTy1MF9Z310HapVHumXKcbphBSzFTbdTLdmUI8mbAtU9gAaplqu7ahO1MYXAC4MVPCQsbQTKEmY/eGa89mZeX5wqrzSYYvLjDzia8mUgBImNN2loUwtBTapmkqkIJmS9NCFtjVUgBY0BQyh5VCk3mUZlLQK+0EGV21CshVK4Ums0G9pMDjUbpIwbY7koIr5iUOLIW2CYUlKSAxTc0wL+ih4zQ1/u2aiAQLc6urRYIcW61IOkxmrLQXoL+uw9VKCuDSNPtmRhFtUxYVSMHUSgoQdhS6AVhoXtyBpdA62XF4KdhaSQF1FW+FyFEqBdxXxKE/KWgWfEK21ZEUXKxWCn1FHHqUgl550BjhbqSAhD0NBpeCHHE4JGb6ZxIz8X8uMfPkopuqxEx8Ip+WnKHNrMmDvm/axLigetrkNc8DbfMztFnyNczkaNwtbQg778yLxPFEomGIk0MwB+IW/xNXyV83tSNOngvL28SgG9wmZjw2Btu2ZNBdYswGGTxtdolhwI3Hp2gcck8l4Ar7WdVsqgR47GigXZXkDqpuV6UprbcXhiScwim4Nd8pBNAJhULmSM1f7BThwGHcp8FiQbtNlsZjPdwHIUvobXRwoLowfDYMYDIcsb0NOZZHt6GoWvSnbMoGDOedD0xh5ZEw+3+SvSkr4TZ8slBVJ7oQxiDqfbLJIGQWr9Yh2U3TeE1xWexvrVktNofsfkhSM4gEfBOBlgTS0+OmnVlQ5bjzKXr8Fw==</diagram></mxfile>
```
[the remark this answers is not in the document]
Comment (ckittl) [reply docs/test/complexTopology.drawio]: This is a file that shows the structure of the test grid, I created and possible derivatives of it. I would suggest to keep it here, at least until we have a proper documentation elsewhere.

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+            .map(NodeInput::getVoltLvl)
+            .collect(Collectors.groupingBy(voltLvl -> voltLvl, Collectors.counting()));
+
+    /* At this point only one voltage level should be apparent */
+    int amountOfVoltLvl = voltageLevelCount.size();
+    if (amountOfVoltLvl > 1)
+      throw new InvalidGridException(
+          "There are "
+              + amountOfVoltLvl
+              + " voltage levels apparent, although only one is expected.");
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java]:
```suggestion
 throw new InvalidGridException(
          "There are "
              + amountOfVoltLvl
              + " voltage levels apparent, although only one is expected. Following voltage levels are present: "
              + voltageLevelCount.keySet().stream()
                  .map(VoltageLevel::toString)
                  .collect(Collectors.joining(", ")));
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+   * Disassembles this grid model into sub grid models and returns a topology of the sub grids as a
+   * directed, immutable graph. The direction points from higher to lower voltage level.
+   *
+   * @param gridName Name of the grid
+   * @param rawGrid Container model of raw grid elements
+   * @param systemParticipants Container model of system participants
+   * @param graphics Container element of graphic elements
+   * @return An immutable, directed graph of sub grid topologies.
+   */
+  public static ImmutableGraph<SubGridContainer> buildSubGridTopology(
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java]:
```suggestion
  public static ImmutableGraph<SubGridContainer> buildGridTopologyGraph(
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+      String gridName,
+      RawGridElements rawGrid,
+      SystemParticipants systemParticipants,
+      GraphicElements graphics) {
+    /* Collect the different sub nets. Through the validation of lines, it is ensured, no calvanically connected grid
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java]:
```suggestion
    /* Collect the different sub nets. Through the validation of lines, it is ensured, that no galvanic connected grid
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+      subGrids.put(
+          subnetNumber,
+          new SubGridContainer(
+              gridName, subnetNumber, rawGridElements, systemParticipantElements, graphicElements));
+    }
+    return subGrids;
+  }
+
+  /**
+   * Build an immutable graph of the galvanically separated sub grid topology
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java]:
```suggestion
   * Build an immutable graph of the galvanic separated sub grid topology
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+  }
+
+  /**
+   * Combines a given collection of sub grid containers to a joint model. If the single models do
+   * not fit together, exceptions are thrown.
+   *
+   * @param subGridContainers Collections of already existing sub grid models
+   * @return A joint model
+   */
+  public static JointGridContainer combineSubGridModels(
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java]: Maybe a `combineToJointGrid` would make more sense?

File: src/main/java/edu/ie3/datamodel/utils/ValidationUtils.java [hunk 9]
```diff
@@ -231,4 +260,172 @@ public static boolean checkMeasurementUnit(MeasurementUnitInput measurementUnit)
...
+
+    systemParticipants.getHeatPumps().forEach(entity -> checkNodeAvailability(entity, nodes));
+
+    systemParticipants.getLoads().forEach(entity -> checkNodeAvailability(entity, nodes));
+
+    systemParticipants.getPvPlants().forEach(entity -> checkNodeAvailability(entity, nodes));
+
+    systemParticipants.getStorages().forEach(entity -> checkNodeAvailability(entity, nodes));
+
+    systemParticipants.getWecPlants().forEach(entity -> checkNodeAvailability(entity, nodes));
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/utils/ValidationUtils.java]:
```suggestion
 systemParticipants.allEntitiesAsList().parallelStream().forEach(entity -> checkNodeAvailability((SystemParticipantInput)entity, nodes));
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java (hunk 7 above)
Comment (ckittl) [reply src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java]: No, this is just not in the dictionary.

File: src/main/java/edu/ie3/datamodel/utils/ValidationUtils.java (hunk 9 above)
Comment (ckittl) [reply src/main/java/edu/ie3/datamodel/utils/ValidationUtils.java]: Doesn't work as `EvcsInput` is no child of `SystemParticipantInput`.

**Comment by ckittl** [comment]
!test

File: docs/uml/ModelContainerConcept.puml [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  - boolean checkSubGridDependencyGraph(ImmutableGraph<SubGridContainer> subnetDependencyGraph)
+}
+GridContainer <|-- JointGridContainer
+
+Class SubGridContainer{
+  - int subnet
+  - VoltageLevel predominantVoltageLevel
+  + JointGridContainer(attributes w\o graph)
+  + JointGridContainer(attributes w\o graph, ImmutableGraph<SubGridContainer> subGridTopologyGraph)
+}
```
Comment (johanneshiry) [inline docs/uml/ModelContainerConcept.puml]: This description does _not_ fit with the real implementation. 

As I don't know which one is correct I cannot resolve it.

**Comment by johanneshiry** [comment]
!test

**Comment by ckittl** [comment]
!test

**Pull request merged**